src/GasEstimator.ts

```diff
@@ -0,0 +1,230 @@
+                metadata.signature
+            );
+    }
+
+    const relayEstimation = await methodToEstimate.estimateGas({
+        from: relayWorkerAddress,
+        gasPrice: relayData.gasPrice
+    });
+
+    const correctedEstimation = applyGasCorrectionFactor(relayEstimation);
+
+    return tokenEstimation.plus(correctedEstimation);
+};
+
+/**
+ * Estimates the max possible gas consumed by relaying a transaction using a linearFit
+ * @param contractInteractor object containing the contractInteractor that is used to interact with the blockchain
+ * @param request request that contains the relayRequest/deployRequest and metadata
+ * @param tokenEstimation gas consumed by the token transfer
+ * @returns gas estimation from the relayTransaction
+ */
+export const linearFitMaxPossibleGasEstimation = async (
+    contractInteractor: ContractInteractor,
+    { request, relayData }: RelayRequest | DeployRequest,
+    tokenEstimation: BigNumber
+): Promise<BigNumber> => {
+    if (isDeployRequest(request)) {
+        throw Error('LinearFit estimation not implemented for deployments');
+    }
+
+    const internalEstimation = await contractInteractor.estimateGas({
+        from: relayData.callForwarder,
+        to: request.to,
+        data: request.data,
+        gasPrice: relayData.gasPrice
+    });
+
+    const estimation = applyInternalCorrection(internalEstimation);
+
+    const relayEstimation = estimateMaxPossibleRelayCallWithLinearFit(
+        estimation.toNumber(),
+        tokenEstimation.toNumber()
+    );
+
+    return BigNumber(relayEstimation);
+};
+
+/**
+ * Estimates the max possible gas consumed by transfering an ERC20 token
+ * @param contractInteractor object containing the contractInteractor that is used to interact with the blockchain
+ * @param request request that contains the relayRequest/deployRequest and metadata
+ * @returns gas estimation from the transfer
+ */
+export const estimateMaxPossibleGasTokenTransfer = async (
+    contractInteractor: ContractInteractor,
+    { request, relayData }: RelayRequest | DeployRequest
+): Promise<BigNumber> => {
+    const deploy = isDeployRequest(request);
+    let tokenEstimation: BigNumber;
+    const gasInRequest = BigNumber(request.tokenGas);
+    if (gasInRequest.gt(0)) {
+        tokenEstimation = BigNumber(request.tokenGas);
+    } else {
+        const erc20: ERC20Token = await contractInteractor.getERC20Token(
+            request.tokenContract
+        );
+
+        const methodToEstimate = await erc20.instance.contract.methods.transfer(
+            relayData.feesReceiver,
+            toWei('1')
+        );
+
+        let caller = relayData.callForwarder;
+        if (deploy) {
+            const { from, recoverer, index } = request as DeployRequestStruct;
+            caller = await contractInteractor.getSmartWalletAddress(
+                caller,
+                from,
+                recoverer,
+                index
+            );
+        }
```

Comment:
```suggestion
        const { from, recoverer, index } = request as DeployRequestStruct;
    const caller = isDeployCall
        ? relayData.callForwarder
        : await contractInteractor.getSmartWalletAddress(
            relayData.callForwarder,
            from,
            recoverer,
            index
        );
```

Reply:
sorry, I thought I've already commented on this yesterday, but obviously didn't.

Reply:
updated